src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToComponentCodeActionResolver.cs

```diff
@@ -78,13 +78,36 @@ internal sealed class ExtractToComponentCodeActionResolver(
         using var _ = StringBuilderPool.GetPooledObject(out var builder);
 
         var syntaxTree = componentDocument.GetSyntaxTree();
-        foreach (var usingDirective in GetUsingsInDocument(syntaxTree))
+
+        // Right now this includes all the usings in the original document.
+        // https://github.com/dotnet/razor/issues/11025 tracks reducing to only the required set.
+        var usingDirectives = syntaxTree.GetUsingDirectives();
+        foreach (var usingDirective in usingDirectives)
+        {
+            builder.AppendLine(usingDirective.ToFullString());
+        }
+
+        // If any using directives were added, add a newline before the extracted content.
+        if (usingDirectives.Length > 0)
         {
-            builder.AppendLine(usingDirective);
+            builder.AppendLine();
         }
 
-        var extractedText = text.GetSubTextString(new TextSpan(actionParams.Start, actionParams.End - actionParams.Start)).Trim();
-        builder.Append(extractedText);
+        var indentation = GetIndentation(actionParams.Start, text);
+        var extractedText = text.GetSubTextString(TextSpan.FromBounds(actionParams.Start, actionParams.End));
+        var lines = extractedText.Split('\n');
```

Comment:
I don't think SourceText has a `Split` method, so probably can't get away from `GetSubTextString`, but I think `extractedText.AsSpan().Split('\n')` should allow `UnindentLine` to be written using spans and avoid a bunch of intermediate strings.

Reply:
Ah, good catch. I was worried about the allocations a bit but figured at least it's in the resolve. I always forget `AsSpan()`